packages/redis/src/RedisStreamSource.ts

```diff
@@ -0,0 +1,233 @@
+import {
+    IInputSource,
+    IRequireInitialization,
+    IDisposable,
+    MessageRef,
+    IComponentContext,
+    DefaultComponentContext,
+    Lifecycle,
+    makeLifecycle,
+    failSpan,
+} from "@walmartlabs/cookie-cutter-core";
+import { Span, Tags, Tracer } from "opentracing";
+
+import { IRedisInputStreamOptions, IRedisClient, redisClient, IRedisMessage } from ".";
+import { RedisOpenTracingTagKeys } from "./RedisClient";
+
+export class RedisStreamSource implements IInputSource, IRequireInitialization, IDisposable {
+    private done: boolean = false;
+    private client: Lifecycle<IRedisClient>;
+    private tracer: Tracer;
+    private spanOperationName: string = "Redis Input Source Client Call";
+
+    constructor(private readonly config: IRedisInputStreamOptions) {
+        this.tracer = DefaultComponentContext.tracer;
+    }
+
+    public async *start(): AsyncIterableIterator<MessageRef> {
+        let messages: IRedisMessage[] = [];
+
+        // On initial start attempt to fetch any messages from this consumers PEL
+        const pendingMessagesForConsumer = await this.getPendingMessagesForConsumer();
+        messages.push(...pendingMessagesForConsumer);
+
+        while (!this.done) {
+            // Attempt to reclaim any PEL messages that have exceeded this.config.idleTimoutMS
+            const pendingMessagesForConsumerGroup = await this.getPendingMessagesForConsumerGroup();
+            messages.push(...pendingMessagesForConsumerGroup);
+
+            // Get any new messages in the consumer group to process
+            const newMessages = await this.getNewMessages();
+            messages.push(...newMessages);
+
+            // Process messages to MessageRefs and yield
+            for (const message of messages) {
+                const span = this.tracer.startSpan(this.spanOperationName);
+
+                this.spanLogAndSetTags(
+                    span,
+                    this.config.db,
+                    this.config.readStream,
+                    this.config.consumerGroup
+                );
+
+                try {
+                    const messageRef = new MessageRef(
+                        { streamId: message.streamId },
+                        message,
+                        span.context()
+                    );
+
+                    messageRef.once("released", async () => {
+                        await this.client.xAck(
```

Comment:
what should be the behavior on error here? for instance if the message handler threw an error processing the message? (that would be part of the args here `async (msg, value, err) => {`). We probably don't want to ack it in error case and fail the span?

CC @chrnola

Reply:
> We probably don't want to ack it in error case and fail the span?

I agree.

Since we're going to be exposing the "dequeue count" to the message handlers as metadata users will be able to decide how to handle messages that fail multiple times.

Reply:
I think this is the best way to approach it as well.